src/block/request.rs

```diff
@@ -0,0 +1,667 @@
+    pub fn parse<M: GuestAddressSpace>(
+        desc_chain: &mut DescriptorChain<M>,
+    ) -> result::Result<Request, Error> {
+        let chain_head = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+        // The head contains the request type which MUST be readable.
+        if chain_head.is_write_only() {
+            return Err(Error::UnexpectedWriteOnlyDescriptor);
+        }
+
+        let request_header = desc_chain
+            .memory()
+            .read_obj::<RequestHeader>(chain_head.addr())
+            .map_err(Error::GuestMemory)?;
+
+        let mut request = Request {
+            request_type: RequestType::from(request_header.request_type),
+            data: Vec::new(),
+            sector: request_header.sector,
+            status_addr: GuestAddress(0),
+        };
+
+        let status_desc;
+        let mut desc = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+
+        if !desc.has_next() {
+            status_desc = desc;
+            // Only flush requests are allowed to skip the data descriptor(s).
+            if request.request_type != RequestType::Flush {
+                return Err(Error::DescriptorChainTooShort);
+            }
+        } else {
+            while desc.has_next() {
+                if desc.is_write_only()
+                    && (request.request_type == RequestType::Out
+                        || request.request_type == RequestType::Discard
+                        || request.request_type == RequestType::WriteZeroes)
+                {
+                    return Err(Error::UnexpectedWriteOnlyDescriptor);
+                }
+                if !desc.is_write_only() && request.request_type == RequestType::In {
+                    return Err(Error::UnexpectedReadOnlyDescriptor);
+                }
+
+                // Check that the address of the data descriptor is valid in guest memory.
+                let _ = desc_chain
+                    .memory()
+                    .checked_offset(desc.addr(), desc.len() as usize)
+                    .ok_or_else(|| {
+                        Error::GuestMemory(GuestMemoryError::InvalidGuestAddress(desc.addr()))
+                    })?;
+
+                request.data.push((desc.addr(), desc.len()));
+                desc = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+            }
+            status_desc = desc;
+        }
+
+        // The status MUST always be writable.
+        if !status_desc.is_write_only() {
+            return Err(Error::UnexpectedReadOnlyDescriptor);
+        }
+        if status_desc.len() < 1 {
+            return Err(Error::DescriptorLengthTooSmall);
+        }
+
+        // Check that the address of the status descriptor is valid in guest memory.
+        // We will write an u32 status here after executing the request.
+        let _ = desc_chain
+            .memory()
+            .checked_offset(status_desc.addr(), mem::size_of::<u32>())
```

Comment:
The Virtio spec defines that the status field is a u8, so should we check "u32" or "u8"?